wicker/core/datasets.py

```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
+        """Get data item at index within arrow table.
+
+        Pulls from either cache or data store the item in the dataset at index specified.
+
+        Args:
+            idx (int): idx in arrow table to grab data.
+
+        Returns:
+            Dict[str, Any]: Row of data defined through schema object.
+        """
+        tbl = self.arrow_table()
+        columns = self._columns_to_load if self._columns_to_load is not None else tbl.column_names
+        row = {col: tbl[col][idx].as_py() for col in columns}
+        schema = self.schema()
+        return dataloading.load_example(
+            self._column_bytes_file_reader.resolve_pointers(row, schema),
+            schema,
         )
 
-        self._dataset_id = DatasetID(name=dataset_name, version=dataset_version)
-        self._partition = DatasetPartition(dataset_id=self._dataset_id, partition=dataset_partition_name)
-        self._dataset_definition = DatasetDefinition(
-            self._dataset_id,
-            schema=self.schema(),
+    def __len__(self) -> int:
+        """Get length of arrow table inferring it is the same as the dataset.
+
+        Returns:
+            int: length of arrow table.
+        """
+        return len(self.arrow_table())
+
+    def _build_column_bytes_reader(self) -> ColumnBytesFileReader:
+        """Builder function to return a column bytes file reader."""
+        if self._local_cache_path_prefix:
+            # If caching column files from storage into the local filesystem, return a cache-enabled reader.
+            return ColumnBytesFileCache(
+                local_cache_path_prefix=self._local_cache_path_prefix,
+                filelock_timeout_seconds=self._filelock_timeout_seconds,
+                path_factory=self._path_factory,
+                storage=self._storage,
+                dataset_name=self._dataset_name,
+            )
+        # Otherwise return a pass-through reader.
+        return ColumnBytesFileReader(
+            dataset_name=self._dataset_name,
+            path_factory=self._path_factory,
         )
-        self.filters = filters
+
+    def arrow_table(self) -> pyarrow.Table:
+        """Grab and load arrow table from expected path.
+
+        Returns:
+            pyarrow.Table: Arrow table object for the loaded dataset.
+        """
+        path = self._path_factory._get_dataset_partition_path(self._partition)
+        if not self._arrow_table:
+            self._arrow_table = papq.read_table(
+                path,
+                columns=self._columns_to_load,
+                filesystem=self._pa_filesystem,
+                filters=self._filters,
+            )
+        return self._arrow_table
 
     def schema(self) -> DatasetSchema:
+        """Return the schema of the dataset."""
         if self._schema is None:
-            schema_path = self._s3_path_factory.get_dataset_schema_path(self._dataset_id)
+            schema_path = self._path_factory._get_dataset_schema_path(self._dataset_id)
```

Comment:
Is the schema_path abs S3 path? I did not see a local file path factory, correct me if I was wrong

Reply:
My point is per my understanding the path should be different between the mounted filesystem and S3 path, correct me if I was wrong.

Reply:
For `FileSystemDataset` we work with the `WickerPathFactory` base class that is basically like local path factory (`S3PathFactory` is a subclass of `WickerPathFactory`).

So in this line, `schema_path` will end up being an absolute local path for `FileSystemDataset` and it will be an absolute S3 path for `S3Dataset`.